src/Microsoft.AspNetCore.SignalR.Redis/Internal/RedisProtocol.cs

```diff
@@ -35,136 +39,181 @@ public RedisProtocol(IReadOnlyList<IHubProtocol> protocols)
-                if (idCount > 0)
+            // Read excluded Ids
+            IReadOnlyList<string> excludedIds = null;
+            var idCount = MsgPackUtil.ReadArrayHeader(ref data);
+            if (idCount > 0)
+            {
+                var ids = new string[idCount];
+                for (var i = 0; i < idCount; i++)
                 {
-                    var ids = new string[idCount];
-                    for (var i = 0; i < idCount; i++)
-                    {
-                        ids[i] = reader.ReadString();
-                    }
-
-                    excludedIds = ids;
+                    ids[i] = MsgPackUtil.ReadString(ref data);
                 }
 
-                var message = SerializedHubMessage.ReadAllSerializedVersions(reader);
-                return new RedisInvocation(message, excludedIds);
+                excludedIds = ids;
             }
+
+            // Read payload
+            var message = ReadSerializedHubMessage(ref data);
+            return new RedisInvocation(message, excludedIds);
         }
 
-        public RedisGroupCommand ReadGroupCommand(byte[] data)
+        public RedisGroupCommand ReadGroupCommand(ReadOnlyMemory<byte> data)
         {
             // See WriteGroupCommand for format.
-            using (var stream = new MemoryStream(data))
-            using (var reader = new BinaryReaderWithVarInt(stream, _utf8NoBom))
-            {
-                var id = reader.ReadVarInt();
-                var serverName = reader.ReadString();
-                var action = (GroupAction)reader.ReadByte();
-                var groupName = reader.ReadString();
-                var connectionId = reader.ReadString();
 
-                return new RedisGroupCommand(id, serverName, action, groupName, connectionId);
+            var length = MsgPackUtil.ReadArrayHeader(ref data);
+
+            if (length < 5)
+            {
+                throw new InvalidDataException("Insufficient items in GroupCommand array.");
             }
+
+            var id = MsgPackUtil.ReadInt32(ref data);
+            var serverName = MsgPackUtil.ReadString(ref data);
+            var action = (GroupAction)MsgPackUtil.ReadByte(ref data);
+            var groupName = MsgPackUtil.ReadString(ref data);
+            var connectionId = MsgPackUtil.ReadString(ref data);
+
+            return new RedisGroupCommand(id, serverName, action, groupName, connectionId);
         }
 
-        public int ReadAck(byte[] data)
+        public int ReadAck(ReadOnlyMemory<byte> data)
         {
             // See WriteAck for format
-            using (var stream = new MemoryStream(data))
-            using (var reader = new BinaryReaderWithVarInt(stream, _utf8NoBom))
+
+            var length = MsgPackUtil.ReadArrayHeader(ref data);
+
+            if (length < 1)
             {
-                return reader.ReadVarInt();
+                throw new InvalidDataException("Insufficient items in Ack array.");
```

Comment:
How about adding a helper method to MsgPackUtil for this check: `MsgPackUtil.ValidateItems(1, "Ack");`

Should be re-usable here and in the hub protocol.

Reply:
I think we should keep it to this type (and have a similar one in the hub protocol) rather than sharing it. We don't really want to share a public or internal type between the two assemblies and they have somewhat different usage.